pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/v311_00_assets"
+)
+
+const (
+	KubeApiserverStaticPodFileName = "kube-apiserver-pod.yaml"
+	RecoveryPodFileName            = "recovery-kube-apiserver-pod.yaml"
+	RecoveryPodNamespace           = "kube-system"
+	RecoveryCofigFileName          = "config.yaml"
+	AdminKubeconfigFileName        = "admin.kubeconfig"
+
+	RecoveryPodAsset    = "v3.11.0/kube-apiserver/recovery-pod.yaml"
+	RecoveryConfigAsset = "v3.11.0/kube-apiserver/recovery-config.yaml"
+)
+
+var (
+	Scheme = runtime.NewScheme()
+	Codecs = serializer.NewCodecFactory(Scheme)
+)
+
+func init() {
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+	utilruntime.Must(corev1.AddToScheme(Scheme))
+}
+
+type Apiserver struct {
+	KubeApiserverImage string
+	PodManifestDir     string
+	ResourceDirPath    string
+	RecoveryDirPath    string
+
+	kubeApiserverStaticPod *corev1.Pod
+	restConfig             *rest.Config
+	kubeClientSet          *kubernetes.Clientset
+}
+
+func (s *Apiserver) GetCurrentApiserverManifest() (*corev1.Pod, error) {
+	if s.kubeApiserverStaticPod != nil {
+		return s.kubeApiserverStaticPod, nil
+	}
+
+	kubeApiserverManifestPath := path.Join(s.PodManifestDir, KubeApiserverStaticPodFileName)
+	f, err := os.OpenFile(kubeApiserverManifestPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file %q: %v", kubeApiserverManifestPath, err)
+	}
+	defer f.Close()
+
+	buf := bytes.NewBuffer(nil)
+	_, err = io.Copy(buf, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %q: %v", kubeApiserverManifestPath, err)
+	}
+
+	obj, err := runtime.Decode(Codecs.UniversalDecoder(corev1.SchemeGroupVersion), buf.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode file %q: %v", kubeApiserverManifestPath, err)
+	}
+
+	// TODO: support conversions if the object is in different but convertible version
+	kubeApiserverStaticPod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("unsupported type: kubeApiserverStaticPodManifest is not type *corev1.Pod but %T", obj)
+	}
+
+	s.kubeApiserverStaticPod = kubeApiserverStaticPod
+
+	return s.kubeApiserverStaticPod, nil
+}
+
+func (s *Apiserver) GetResourceDirPath() (string, error) {
+	if s.ResourceDirPath != "" {
+		return s.ResourceDirPath, nil
+	}
+
+	pod, err := s.GetCurrentApiserverManifest()
+	if err != nil {
+		return "", fmt.Errorf("failed to read apiserver pod manifest: %v", err)
+	}
+
+	resourceDir, err := getVolumeHostPathPath("resource-dir", pod.Spec.Volumes)
+	if err != nil {
+		return "", fmt.Errorf("failed to find resource-dir: %v", err)
+	}
+
+	s.ResourceDirPath = *resourceDir
+
+	return s.ResourceDirPath, nil
+}
+
+func (s *Apiserver) GetRecoveryDirPath() (string, error) {
+	if s.RecoveryDirPath != "" {
+		return s.RecoveryDirPath, nil
+	}
+
+	// Create tmp dir for recovery apiserver
+	recoveryDir, err := ioutil.TempDir("", "kube-recovery-apiserver")
```

Comment:
anybody cleaning this up? If not, add a comment that we accept leaking temp dirs.

Reply:
This will contain secrets. Better make it 0700.

Reply:
it contains both configmaps and secrets so technicaly the dir is 755 and the secret file need to be 600. it is the same as in /etc where others can read the dir and configmaps, not the secret files. we have to make sure secrets have 600 though

Reply:
we are now using /etc/kubernetes/manifests and destroying after